text/0000-pack-phase-debug-script.md

```diff
@@ -0,0 +1,114 @@
+# Meta
+[meta]: #meta
+- Name: Pack phase debug script
+- Start Date: 2025-04-04
+- Author(s): BarDweller
+- Status: Draft 
+- RFC Pull Request:
+- CNB Pull Request: 
+- CNB Issue:
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+An ability to dynamically add and invoke a debug script instead of the lifecycle binary per phase during `pack build`.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Debugging state within a builder can be tricky, it can be beneficial during buildpack and platform development to understand the content of buildpack toml files, and filesystem permissions, and even know sha hashes for builder packaged content. While it is possible to test some things via a custom debug buildpack, buildpacks are not involved in every phase of execution, and some changes occur outside of their involvement. Being able to trace those state changes and verify assumptions is beneficial to buildpack and platform developers alike. 
+- What use cases does it support?
+    - Example scenarios: 
+      - Trying to validate filesystem ownership/permissions/layout within a build container
+      - Validating cnb toml content pre-post phase execution to ensure changes are as expected
+      - Dumping checksum hashes of builder content to validate binaries being tested are the right ones
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The pack cli would allow specification of a debug script, that should be driven instead of the appropriate lifecycle binary for each phase. 
+
+The script will be passed the name/path and arguments of the lifecycle binary that would have been driven. 
+
+The script is able to run abitrary commands and dump information to stdout/stderr, these commands can include invoking the original lifecycle binary with its original arguments. 
+
+A simple debug script could look like.. 
```

Comment:
Can this be any executable? Or does it have to be a shell script?

Reply:
I'd imagine in practice this will be whatever Go can exec, in the same manner as the build/detect scripts/binaries.